pallets/sgx-account-linker/src/lib.rs

```diff
@@ -43,172 +43,366 @@ type Signature = [u8; 65];
-			layer_one_block_number: T::BlockNumber,
-			expiring_block_number: T::BlockNumber,
-			sig: Signature,
-		) -> DispatchResultWithPostInfo {
-
-			// in sgx runtime, the account who want to link ethereum address don't have the balance to
-			// submit extrinsic, the origin could be the root account
-			let _ = ensure_signed(origin)?;
-			Self::do_link_eth(account,
-				index,
-				addr_expected,
-				expiring_block_number,
-				layer_one_block_number,
-				sig)
-		}
-	}
-
-	impl<T:Config> Pallet<T> {
-		/// Assemble the message that the user has signed 
-		/// Format: "Link Litentry: " + Litentry account + expiring block number
-		fn generate_raw_message(account: &T::AccountId, expiring_block_number: T::BlockNumber) -> Vec<u8> {
-			let mut bytes = b"Link Litentry: ".encode();
-			let mut account_vec = account.encode();
-			let mut expiring_block_number_vec = expiring_block_number.encode();
-			
-			bytes.append(&mut account_vec);
-			bytes.append(&mut expiring_block_number_vec);
-			bytes
-		}
-
-		pub fn do_link_eth(
-			account: T::AccountId,
-			index: u32,
-			addr_expected: EthAddress,
-			expiring_block_number: T::BlockNumber,
-			layer_one_blocknumber: T::BlockNumber,
-			sig: Signature,
-		) -> DispatchResultWithPostInfo {
-			ensure!(expiring_block_number > layer_one_blocknumber, Error::<T>::LinkRequestExpired);
-			ensure!((expiring_block_number - layer_one_blocknumber) < T::BlockNumber::from(EXPIRING_BLOCK_NUMBER_MAX),
-				Error::<T>::InvalidExpiringBlockNumber);
-
-			let bytes = Self::generate_raw_message(&account, expiring_block_number);
-
-			let hash = util_eth::eth_data_hash(bytes).map_err(|_| Error::<T>::UnexpectedEthMsgLength)?;
-
-			let mut msg = [0u8; 32];
-			msg[..32].copy_from_slice(&hash[..32]);
-
-			let addr = util_eth::addr_from_sig(msg, sig)
-				.map_err(|_| Error::<T>::EcdsaRecoverFailure)?;
-			ensure!(addr == addr_expected, Error::<T>::UnexpectedAddress);
-
-			EthereumLink::<T>::mutate(&account, |addrs| {
-				let index = index as usize;
-				// NOTE: allow linking `MAX_ETH_LINKS` eth addresses.
-				if (index >= addrs.len()) && (addrs.len() != MAX_ETH_LINKS) {
-					addrs.push(addr.clone());
-				} else if (index >= addrs.len()) && (addrs.len() == MAX_ETH_LINKS) {
-					addrs[MAX_ETH_LINKS - 1] = addr.clone();
-				} else {
-					addrs[index] = addr.clone();
-				}
-			});
-
-			Self::deposit_event(Event::EthAddressLinked(account, addr.to_vec()));
-
-			Ok(().into())
-
-		}
-	}
+    use crate::*;
+    use codec::Encode;
+    use frame_support::{dispatch::DispatchResultWithPostInfo, pallet_prelude::*};
+    use frame_system::ensure_signed;
+    use frame_system::pallet_prelude::*;
+    use sp_core::{ed25519, sr25519};
+    use sp_std::prelude::*;
+
+    use weights::WeightInfo;
+    pub const EXPIRING_BLOCK_NUMBER_MAX: u32 = 10 * 60 * 24 * 30; // 30 days for 6s per block
+    pub const MAX_ETH_LINKS: usize = 3;
+    pub const MAX_BTC_LINKS: usize = 3;
+    pub const MAX_POLKADOT_LINKS: usize = 3;
```

Comment:
we may rename this and the storage from polkadot link to substrate link, to avoid misleading.

Reply:
accepted.